crypto/ct/ct_oct.c

```diff
@@ -0,0 +1,570 @@
+                          *((c)++)=(unsigned char)(((l)    )&0xff))
+
+static STACK_OF(SCT) *d2i_SCT_LIST(STACK_OF(SCT) **a,
+                                   const unsigned char **pp, int len);
+static int i2d_SCT_LIST(STACK_OF(SCT) *a, unsigned char **pp);
+static int i2r_SCT_LIST(X509V3_EXT_METHOD *method, STACK_OF(SCT) *sct_list,
+                        BIO *out, int indent);
+
+static char *i2s_poison(const X509V3_EXT_METHOD *method, void *val)
+{
+    return OPENSSL_strdup("NULL");
+}
+
+const X509V3_EXT_METHOD v3_ct_scts[] = {
+    { NID_ct_precert_scts, 0, NULL,
+    0, (X509V3_EXT_FREE)SCT_LIST_free,
+    (X509V3_EXT_D2I)d2i_SCT_LIST, (X509V3_EXT_I2D)i2d_SCT_LIST,
+    0, 0, 0, 0,
+    (X509V3_EXT_I2R)i2r_SCT_LIST, 0,
+    NULL },
+
+    { NID_ct_precert_poison, 0, ASN1_ITEM_ref(ASN1_NULL),
+    0, 0, 0, 0, i2s_poison, 0,
+    0, 0, 0, 0, NULL },
+
+    { NID_ct_cert_scts, 0, NULL,
+    0, (X509V3_EXT_FREE)SCT_LIST_free,
+    (X509V3_EXT_D2I)d2i_SCT_LIST, (X509V3_EXT_I2D)i2d_SCT_LIST,
+    0, 0, 0, 0,
+    (X509V3_EXT_I2R)i2r_SCT_LIST, 0,
+    NULL },
+};
+
+int SCT_signature_is_valid(const SCT *sct)
+{
+    if (sct == NULL) {
+        CTerr(CT_F_SCT_SIGNATURE_IS_VALID, ERR_R_PASSED_NULL_PARAMETER);
+        return -1;
+    }
+
+    if (SCT_get_signature_nid(sct) == NID_undef ||
+        sct->sig_len == 0 || sct->sig == NULL) {
+        return 0;
+    }
+
+    return 1;
+}
+
+
+int o2i_SCT_signature(SCT *sct, const unsigned char **in, size_t len)
+{
+    int ret = -1;
+    size_t siglen;
+    size_t len_remaining = len;
+    const unsigned char *p = *in;
+
+    if (sct == NULL || in == NULL || *in == NULL) {
+        CTerr(CT_F_O2I_SCT_SIGNATURE, ERR_R_PASSED_NULL_PARAMETER);
+        goto err;
+    }
+
+    if (sct->version == 0) { /* SCT v1 */
+        /*
+         * digitally-signed struct header: (1 byte) Hash algorithm (1 byte)
+         * Signature algorithm (2 bytes + ?) Signature
+         *
+         * This explicitly rejects empty signatures: they're invalid for
+         * all supported algorithms.
+         */
+        if (len <= 4) {
+            CTerr(CT_F_O2I_SCT_SIGNATURE, CT_R_SCT_INVALID_SIGNATURE);
+            goto err;
+        }
+
+        /* Get hash and signature algorithm */
+        sct->hash_alg = *p++;
+        sct->sig_alg = *p++;
+        if (SCT_get_signature_nid(sct) == NID_undef) {
+            CTerr(CT_F_O2I_SCT_SIGNATURE, CT_R_SCT_INVALID_SIGNATURE);
+            goto err;
+        }
+        /* Retrieve signature and check it is consistent with the buffer length */
+        n2s(p, siglen);
+        len_remaining -= (p - *in);
+        if (siglen > len_remaining) {
+            CTerr(CT_F_O2I_SCT_SIGNATURE, CT_R_SCT_INVALID_SIGNATURE);
+            goto err;
+        }
+
+        if (SCT_set1_signature(sct, p, siglen) != 1)
+            goto err;
+        len_remaining -= siglen;
+
+        *in = p + siglen;
+        ret = len - len_remaining;
+    } else {
```

Comment:
I prefer to see the test inverted with the early goto err and then the main body gets undented one level and ends with a return.